File: 10/JYC/2360. Longest Cycle in a Graph.ts
function longestCycle(edges: number[]): number {
  const visited = new Array(edges.length).fill(false);
  let result = -1;

  const DFS = (n: number, depth: number, register: Record<number, number>) => {
    if (register[n] != null) {
      result = Math.max(depth - register[n], result);
    } else {
      if (!visited[n]) {
        visited[n] = true;
        const next = edges[n];
        if (next != -1) {
          register[n] = depth;
          DFS(next, depth + 1, register);
        }
      }
    }
  };

  for (let i = 0; i < edges.length; i++) {
    if (!visited[i]) {
      DFS(i, 0, {});
    }
  }

  return result;
}

File: 10/haechan/2360 Longest Cycle in a Graph.kt
import java.util.*
import kotlin.math.*

class Solution {
    lateinit var stack: Stack<Int>

    fun longestCycle(edges: IntArray): Int {
        val visited = BooleanArray(edges.size) { false }
        var longestCycle = -1
        
        for (i in edges.indices) {
            stack = Stack<Int>().apply {
                push(i)
            }

            while (stack.isNotEmpty()) {
                val current = stack.peek()

                visited[current] = true

                val next = edges[current]
                if (next == -1) break

                if (visited[next]) {
                    if (stack.contains(next)) {
                        var count = 1
                        while (stack.pop() != next) {
                            count++
                        }
                        longestCycle = max(longestCycle, count)
                    }
                    break
                }

                stack.push(next)
            }
        }

        return longestCycle
    }
}
